src/geometry/contacts.rs

```diff
@@ -343,13 +345,24 @@ fn compute_contacts_for_pair_of_cells(
                     let pj = if is_boundary_j {
                         let bj = &boundaries[fluid_j];
                         if !fluids[*fluid_i]
-                            .boundary_interaction
-                            .test(bj.fluid_interaction)
+                            .interaction_groups
+                            .test(bj.interaction_groups)
                         {
                             continue;
                         }
                         boundaries[fluid_j].positions[particle_j]
                     } else {
+                        if *fluid_i != fluid_j {
+                            let fluid_i_data = &fluids[*fluid_i];
+                            let fluid_j_data = &fluids[fluid_j];
+
+                            if !fluid_i_data
+                                .interaction_groups
+                                .test(fluid_j_data.interaction_groups)
+                            {
+                                continue;
+                            }
```

Comment:
Not very important. Just aesthetic preferences:
```suggestion
                            let groups_i = &fluids[*fluid_i].interaction_groups;
                            let groups_j = &fluids[fluid_j].interaction_groups;

                            if !groups_i.test(groups_j) {
                                continue;
                            }
```